SILOptimizer: a new small optimization pass to remove redundant basic block arguments.

eeckstein opened the pull request:
The BlockArgumentOptPass eliminates block arguments which have the same value as other arguments of the same block.
This also works with cycles, like two equivalent loop induction variables. Such patterns are generated e.g. when using stdlib's enumerated() on Array.

```
preheader:
  br bb1(%initval, %initval)
header(%phi1, %phi2):
  %next1 = builtin "add" (%phi1, %one)
  %next2 = builtin "add" (%phi2, %one)
  cond_br %loopcond, header(%next1, %next2), exit
exit:
```
is replaced with
```
preheader:
  br bb1(%initval)
header(%phi1):
  %next1 = builtin "add" (%phi1, %one)
  %next2 = builtin "add" (%phi1, %one) // dead: will be cleaned-up later
  cond_br %loopcond, header(%next1), exit
exit:
```

Any remaining dead or "trivially" equivalent instructions will then be cleaned-up by DCE and CSE, respectively.

rdar://problem/33438123

eeckstein committed:
SILOptimizer: a new small optimization pass to remove redundant basic block arguments.

RedundantPhiElimination eliminates block phi-arguments which have the same value as other arguments of the same block.
This also works with cycles, like two equivalent loop induction variables. Such patterns are generated e.g. when using stdlib's enumerated() on Array.

   preheader:
     br bb1(%initval, %initval)
   header(%phi1, %phi2):
     %next1 = builtin "add" (%phi1, %one)
     %next2 = builtin "add" (%phi2, %one)
     cond_br %loopcond, header(%next1, %next2), exit
   exit:

is replaced with

   preheader:
     br bb1(%initval)
   header(%phi1):
     %next1 = builtin "add" (%phi1, %one)
     %next2 = builtin "add" (%phi1, %one) // dead: will be cleaned-up later
     cond_br %loopcond, header(%next1), exit
   exit:

Any remaining dead or "trivially" equivalent instructions will then be cleaned-up by DCE and CSE, respectively.

rdar://problem/33438123

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci benchmark

gottesmm commented:
@eeckstein Not reviewing this, but curious, doesn't simplify cfg do this? Why do we need an extra pass. Are you hoping to remove that code from simplify-cfg?

eeckstein commented:
@gottesmm No, SimplifyCFG does not perform this optimization. I didn't want include this into SimplifyCFG, because

1. It doesn't need to. No other SimplifyCFG optimization depends on this.
2. SimplifyCFG is already very complex
3. Compile time: this optimization doesn't need to run every time SimplifyCFG runs.

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci benchmark

atrick reviewed (changes requested):
We need to start documenting pass order motivations, expectations, and dependencies. I think the file-level comment should always have a small section at the top specifying which passes should run before/after and any motivations for running the pass at a particular point. In addition it's sometimes helpful to comment the pass pipeline itself, which nobody does.

In particuler, it's critical to point out that this pass will repeat most of its work each time it runs until dead phis are eliminated. So I guess there's a fairly hard dependency on CSE running in between.

---

It's good to have this optimization. This is a variant of graph isomorphism where you can label the phi nodes and their operands. But it can still blow up in pathological cases. The problem with the straightforward worklist is that it's actually O(n^4), not quadratic (blocks^2 * arguments^2). With the threshold you added, it's now quadratic in the number of nodes in a cycle (times 100 combinations of arguments per block).

When I implemented this in the past, I made sure to process the phi strongly-connected components only once as a unit so it ends up being linear.

I think you might be able to avoid that if the block order that you visit happens to be RPO. This way, the upward-adjacent phi node in cycle will already have been processed. But if the algorithmic complexity depends on that, it shouldn't just be assumed. You should probably explicitly visit in RPO order and comment why that's necessary. Or, if you think forward block order is a good enough approximation to RPO, you could just comment that you're making that assumption.

You would only be able to actually evaluate this issue if you auto-generate large phi cycles and see how it scales with different block orders.

gottesmm commented:
@eeckstein my thought was maybe we could use this to remove that functionality from simplify cfg.

gottesmm commented:
and thus simplify simplify cfg.

eeckstein commented:
@gottesmm As I said, there is no such optimization in SimplifyCFG

eeckstein commented:
@atrick Thanks for the review! I added some comments.

About complexity: you are right that in pathological cases it can blow up. Even worse, it can also happen with no cycles: in case there are very long chains of side-effect-free instructions.

So I just added a limit for the worklist size as well. I made some experiments with the benchmarks and chose the limit accordingly.

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci smoke test

atrick commented:
@eeckstein Good, I'm glad you thought about the pathological case. You're right, this is a problem even if the phi chains don't end up being redundant, so just processing it in RPO order won't solve it. I just know I had to be very careful when implementing this in the past. But in my case, I wanted to guarantee all redundant phis were removed for verification purposes. If you just want to weed out the cases that happen in practice, a limit on the worklist should be fine.

atrick reviewed (approved):
Thanks

atrick reviewed (commented):
The name "CommonArgumentElimination" is extremely misleading to me. I think it should be "RedundantPhiElimination".

I also suggest using the `isPhiArgument` API, which both cuts down on the work that this is doing and makes the intention clear. When we finally fix the type name `SILPhiArgument` to only apply to phis, then it should be obvious how to adjust passes like this.

eeckstein commented:
Yeah, makes sense. I renamed it

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci smoke test

compnerd commented:
@eeckstein I think that this change might have triggered a regression in the Foundation build with optimizations.  I'm going to try to isolate a test case and see if this is truly the case, but perhaps you can spot it faster than I can do that.  The failure logs are at https://dev.azure.com/compnerd/3133d6ab-80a8-4996-ac4f-03df25cd3224/_apis/build/builds/28960/logs/82

eeckstein commented:
@compnerd Hm, this verification failure doesn't look like it's related. Maybe the optimization uncovered another bug.
Can you run the compiler with -Xfrontend -sil-verify-all? It should give you the optimization pass which really causes the problem.
